float32   w
float32   x
float32   y
float32   z